hedyorg/hedy
# 💻 Create student accounts in bulk

This PR redesigns the screen that allows teachers to create multiple student accounts for a class. As discussed during the contributors' meetings the page should allow teachers to:
- create students and let Hedy auto generate passwords
- create students and provide their own passwords using a semicolon as a separator
- paste a two-column spreadsheet and expect Hedy to transform the input accordingly
- copy to clipboard and print the successfully created accounts

Fixes #3936 

Things to note
- When an error occurs, the error message does not go away until you remove it through its close button. This is done intentionally because the error message could point out which rows of your input are incorrect. So, if it just goes away after x seconds, I think this could be very inconvenient.
- The loading indicator which is displayed when the accounts are being created should take the whole page, because we do not want people to go away from the page while accounts are being created. The loading indicator is currently in the layout.html as an attempt to make it reusable but maybe when it takes the whole page, it will not be reusable.

**How to test**
Features to test:
- You should be able to create one or multiple student accounts with and without passwords.
- When you print from Excel, the separators should be automatically changed to semicolons. Note that this should work if you use Ctrl/Cmd + V or right-click and chose Paste from the context menu.
- You should also be able to successfully create student accounts when you are a second teacher. In the devdb teacher4 is a second teacher of CLASS1.
- You should be able to print and copy to clipboard the successfully created accounts. Note that sometimes when you try to print the accounts information, the borders of the table will not be shown in the print-preview screen. I do not know why this happens, but the they are correctly printed. I use 'Save to PDF' functionality and the pdf files look great.

Errors to test:
- You forgot to put in any usernames.
- You tried to create usernames with invalid characters. Currently these are `@` and `:`.
- You got the auto generate passwords toggle checked, but you provided usernames and passwords correctly separated with semicolon. In this case you should get a specific error that directs you to the toggle. 
- You got the auto generate passwords toggle unchecked, but you provided only usernames and no passwords. In this case you should get a nice error directing you to the toggle.
- You got the auto generate passwords toggle unchecked and you provide multiple usernames and passwords. However, only 1 row is missing a semicolon. You should get an error that directs you to this row.
- You got the auto generate passwords toggle unchecked but some of your passwords are less than 6 characters.
- You supplied the same username twice, with auto generated passwords on and off.
- You supplied a username that is already stored in the database. You should get an error and a way to resolve it.
- You tried to create more than 100 usernames.
